server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/service/ExpireInboxesService.java

```diff
@@ -0,0 +1,209 @@
+            SearchQuery expiration = SearchQuery.of(
+                runningOptions.maxAgeDuration.map(maxAge -> {
+                        Date limit = Date.from(now.toInstant().minus(maxAge));
+                        return SearchQuery.internalDateBefore(limit, DateResolution.Second);
+                    })
+                    .orElse(
+                        SearchQuery.headerDateBefore("Expires", now, DateResolution.Second)
+                    )
+            );
+            return Iterators.toFlux(usersRepository.list())
+                .transform(ReactorUtils.<Username, Task.Result>throttle()
+                    .elements(runningOptions.getUsersPerSecond())
+                    .per(Duration.ofSeconds(1))
+                    .forOperation(username -> expireUserInbox(context, username, expiration)))
+                .reduce(Task.Result.COMPLETED, Task::combine);
+        } catch (UsersRepositoryException e) {
+            LOGGER.error("Error while accessing users from repository", e);
+            return Mono.just(Task.Result.PARTIAL);
+        }
+    }
+
+    private Mono<Result> expireUserInbox(Context context, Username username, SearchQuery expiration) {
+        MailboxSession session = mailboxManager.createSystemSession(username);
+        MailboxPath mailboxPath = MailboxPath.inbox(username);
+        return Mono.from(mailboxManager.getMailboxReactive(mailboxPath, session))
+            // newly created users do not have mailboxes yet, just skip them
+            .onErrorResume(MailboxNotFoundException.class, ignore -> Mono.empty())
+            .flatMap(mgr -> searchMessagesReactive(mgr, session, expiration)
+                .flatMap(list -> deleteMessagesReactive(mgr, session, list)))
+            .doOnNext(expired -> {
+                if (expired) {
+                    context.incrementExpiredCount();
+                }
+                context.incrementProcessedCount();
+            })
+            .then(Mono.just(Task.Result.COMPLETED))
+            .onErrorResume(e -> {
+                LOGGER.warn("Failed to expire user mailbox {}", username, e);
+                context.incrementFailedCount();
+                context.incrementProcessedCount();
+                return Mono.just(Task.Result.PARTIAL);
+            });
+    }
+
+    private Mono<List<MessageUid>> searchMessagesReactive(MessageManager mgr, MailboxSession session, SearchQuery expiration) {
+        try {
+            return Flux.from(mgr.search(expiration, session)).collectList();
+        } catch (MailboxException e) {
+            return Mono.error(e);
+        }
+    }
+
+    private Mono<Boolean> deleteMessagesReactive(MessageManager mgr, MailboxSession session, List<MessageUid> uids) {
+        try {
+            if (uids.isEmpty()) {
+                return Mono.just(false);
+            } else {
+                mgr.delete(uids, session);
+                return Mono.just(true);
```

Comment:
Warning! A blocking call in a reactive pipeline!

This will cause a major performance collapse.

As a rule of thumbs blocking calls needs to be at the very least wrapped:

```
Mono.fromRunnable(() -> blocking()).subscribeOn(elastic());
```

(Which is better than nothing but still can spawn dozens of threads blocking to wait other reactor tasks to complete... So still wasteful!)

Here we could very easily expose a `deleteReactive` method to not even be blocking!

Reply:
I do not feel comfortable in messing with the MessageManager, so I will go with the wrap approach. Feel free to change this later at your convenience :)

Reply:
Ok but at the very least wrap the blocking call for now ;-)

I'd be happy to get one more reactify-everything-commit on my kill list.